src/plugins/chatMarkers/ChatMarkersPlugin.ts

```diff
@@ -0,0 +1,410 @@
+import { AbstractPlugin } from '@src/plugin/AbstractPlugin'
+import PluginAPI from '@src/plugin/PluginAPI'
+import Message from '@src/Message'
+import { IMessage, DIRECTION } from '@src/Message.interface'
+import Contact from '@src/Contact'
+import JID from '@src/JID'
+import Translation from '@src/util/Translation'
+import * as Namespace from '@src/connection/xmpp/namespace'
+import { $msg } from '@src/vendor/Strophe'
+import ChatWindow from '@src/ui/ChatWindow'
+import ChatWindowMessage from '@src/ui/ChatWindowMessage'
+import PersistentMap from '@src/util/PersistentMap'
+
+/**
+ * XEP-0333: Chat Markers
+ *
+ * @version 1.0
+ * @see http://xmpp.org/extensions/xep-0333.html
+ */
+
+const MIN_VERSION = '4.0.0';
+const MAX_VERSION = '4.0.0';
+
+// frequently used strings
+const CHATMARKERS = 'CHATMARKERS';
+const MARKABLE = 'markable';
+const RECEIVED = 'received';
+const DISPLAYED = 'displayed';
+const ACKNOWLEDGED = 'acknowledged';
+const ID = 'id';
+const FROM = 'from';
+
+export default class ChatMarkersPlugin extends AbstractPlugin {
+   private static readonly INDICATOR_TEMPLATE = require('../../../template/plugins/chatMarkers/chatmarkers-indicator.hbs');
+   private static readonly  ACKNOWLEDGE_TEMPLATE = require('../../../template/plugins/chatMarkers/chatmarkers-acknowledge.hbs');
+
+   protected pluginAPI = this.pluginAPI as PluginAPI;
+
+   private data: PersistentMap;
+
+   public static getName(): string {
+      return 'Chat Markers';
+   }
+
+   public static getDescription(): string {
+      return Translation.t('chatmarkers-description');
+   }
+
+   constructor(pluginAPI: PluginAPI) {
+      super(MIN_VERSION, MAX_VERSION, pluginAPI);
+
+      Namespace.register(CHATMARKERS, 'urn:xmpp:chat-markers:0');
+
+      this.pluginAPI.addFeature(Namespace.get(CHATMARKERS));
+
+      this.data = new PersistentMap(this.pluginAPI.getStorage(), 'chatmarkers');
+
+      this.pluginAPI.addPreSendMessageStanzaProcessor(this.preSendMessageStanzaProcessor);
+      this.pluginAPI.addPreSendMessageProcessor(this.preSendMessageProcessor);
+      this.pluginAPI.addAfterReceiveMessageProcessor(this.afterReceiveMessageProcessor);
+
+      this.pluginAPI.getConnection().registerHandler(this.onMarkableMessage, null, 'message');
+      this.pluginAPI.getConnection().registerHandler(this.onChatMarkersMessage, null, 'message');
+
+      this.pluginAPI.registerChatWindowInitializedHook((chatWindow: ChatWindow, contact: Contact) => {
+         this.chatWindowInitializedHook(chatWindow, contact);
+      });
+
+      this.data.registerHook('lastReceivedMsgId', (lastReceivedMsgId: string) => {
```

Comment:
This hook is executed in every tab, which results in multiple messages. Why are you using hooks anyway?